modules/validations/crossing_ways.js

```diff
@@ -39,10 +40,104 @@ export function validationHighwayCrossingOtherWays() {
         return n1 > nA ? n1 + n2 + nA + nB : nA + nB + n1 + n2;
     }
 
+    function getWayAndParentRelationTags(way, graph) {
+        var tags = _clone(way.tags);
+        var parentRels = graph.parentRelations(way);
+        for (var i = 0; i < parentRels.length; i++) {
+            var rel = parentRels[i];
+            for (k in rel.tags) {
+                if (!tags[k]) tags[k] = rel.tags[k];
+            }
+        }
+        return tags;
+    }
+
+    function hasTag(tags, key) {
+        return tags[key] !== undefined && tags[key] !== 'no';
+    }
+
+    function getFeatureTypeForCrossingCheck(way, graph) {
+        var tags = getWayAndParentRelationTags(way, graph);
+        if (hasTag(tags, 'highway')) return 'highway';
+        if (hasTag(tags, 'building')) return 'building';
+        if (hasTag(tags, 'railway')) return 'railway';
+        if (hasTag(tags, 'waterway') || tags.natural === 'water') return 'water';
+
+        return null;
+    }
+
+    function extendTagsByInferredLayer(tags, way) {
+        if (!hasTag(tags, 'layer')) {
+            tags.layer = way.layer().toString();
+        }
+        return tags;
+    }
+
+    function isLegitCrossing(way1, featureType1, way2, featureType2, graph) {
```

Comment:
@bhousel @quincylvania please take a look at all the cases here and comment whichever seems incorrect.